### ISSUE TITLE
Error on plugin upgrade (Trac #67)

### ISSUE THREAD

**Issue by cfobel** [wheeler-microfluidics/microdrop#71]
When attempting to upgrade a plugin with a newer version, I get the following error (Windows 7, running from python interpreter or exe, version 0.1.397-g8647800):

(<type 'exceptions.WindowsError'>, WindowsError(183, 'Cannot create a file when that file already exists'), <traceback object at 0x0621EF08>) {}
  File "C:\Users\ryan\dev\python\microdrop\microdrop\gui\plugin_manager_dialog.py", line 212, in on_button_install_clicked
    self.verify_new_plugin(temp_dir)
  File "C:\Users\ryan\dev\python\microdrop\microdrop\gui\plugin_manager_dialog.py", line 262, in verify_new_plugin
    self.install_plugin(plugin_root, installed_plugin_path)
  File "C:\Users\ryan\dev\python\microdrop\microdrop\gui\plugin_manager_dialog.py", line 292, in install_plugin
    ignore=ignore_patterns('*.pyc'))
  File "C:\Python27\lib\shutil.py", line 174, in copytree
    os.makedirs(dst)
  File "C:\Python27\lib\os.py", line 157, in makedirs
    mkdir(name, mode)

Migrated from http://microfluidics.utoronto.ca/ticket/67
```json
{
    "status": "closed", 
    "changetime": "2014-04-17T19:39:01", 
    "description": "When attempting to upgrade a plugin with a newer version, I get the following error (Windows 7, running from python interpreter or exe, version 0.1.397-g8647800):\n\n(<type 'exceptions.WindowsError'>, WindowsError(183, 'Cannot create a file when that file already exists'), <traceback object at 0x0621EF08>) {}\n  File \"C:\\Users\\ryan\\dev\\python\\microdrop\\microdrop\\gui\\plugin_manager_dialog.py\", line 212, in on_button_install_clicked\n    self.verify_new_plugin(temp_dir)\n  File \"C:\\Users\\ryan\\dev\\python\\microdrop\\microdrop\\gui\\plugin_manager_dialog.py\", line 262, in verify_new_plugin\n    self.install_plugin(plugin_root, installed_plugin_path)\n  File \"C:\\Users\\ryan\\dev\\python\\microdrop\\microdrop\\gui\\plugin_manager_dialog.py\", line 292, in install_plugin\n    ignore=ignore_patterns('*.pyc'))\n  File \"C:\\Python27\\lib\\shutil.py\", line 174, in copytree\n    os.makedirs(dst)\n  File \"C:\\Python27\\lib\\os.py\", line 157, in makedirs\n    mkdir(name, mode)", 
    "reporter": "ryan", 
    "cc": "", 
    "resolution": "fixed", 
    "_ts": "1397763541728826", 
    "component": "microdrop", 
    "summary": "Error on plugin upgrade", 
    "priority": "major", 
    "keywords": "", 
    "version": "0.1", 
    "time": "2012-03-14T02:34:18", 
    "milestone": "Microdrop 1.0", 
    "owner": "cfobel", 
    "type": "defect"
}
```

**Comment by cfobel** [comment]
Trac update at 20120315T05:13:22: ryan changed status from "new" to "assigned"

**Comment by cfobel** [comment]
Trac update at 20120403T11:40:23: 
* christian@fobel.net commented:

> commit d8dec69e675ee3ce1c64dd6f0ed293b4e0eb820e
> Author: Christian Fobel <christian@fobel.net>
> Date:   Tue Apr 3 07:33:55 2012 -0400
> 
>     When upgrading a plugin, install new version to...
>     
>     When upgrading a plugin, install new version to new path.  Use the original
>     plugin path and increment a count to append onto the end until a unique
>     directory name is found.  This allows for the new version of a plugin to be
>     installed before the existing plugin is deleted.  This behaviour is required
>     since the deletion of the existing plugin must be delayed until the program is
>     restarted (due to file locking).
>     
>     Fixes #71 (confirmed on Ubuntu, need to test in Windows)


* christian@fobel.net changed resolution from "" to "fixed"
* christian@fobel.net changed status from "assigned" to "closed"

**Comment by cfobel** [comment]
Trac update at 20120403T14:52:12: 
* ryan commented:

> I would like to maintain the same plugin directory after upgrading, so maybe on restart the program could trigger a rename?


* ryan changed resolution from "fixed" to ""
* ryan changed status from "closed" to "reopened"

**Comment by cfobel** [comment]
Trac update at 20120403T20:15:05: 
* christian@fobel.net commented:

> commit 3d0b7b9789b8626fe3fd78046c4ddda77ac5a538
> Author: Christian Fobel <christian@fobel.net>
> Date:   Tue Apr 3 16:08:02 2012 -0400
> 
>     Process rename queue before loading plugins
>     
>     Process rename queue before loading managed plugins.  This is used to upgrade
>     an existing plugin:
>         -first, the new version is extracted to an available directory name under
>                 the plugins directory
>         -second, the directory containing the stale version of the plugin is marked
>                 for deferred deletion (on next program start)
>         -lastly, the directory containing the new version of the plugin is marked
>                 to be renamed to the original plugin directory name (on next program
>                 start)
>     
>     Fixes #71.


* christian@fobel.net changed resolution from "" to "fixed"
* christian@fobel.net changed status from "reopened" to "closed"

**Comment by cfobel** [comment]
Trac update at 20140417T19:39:01: ryan changed component from "plugins" to "microdrop"
